### PR TITLE
refactored chess pieces code

### PR DESCRIPTION
**Pull request by kevinglezgmz**
Refactored the code for the chess pieces and added some useful functions inside the abstract ChessPiece class

### CODE REVIEW

File: Chess/ChessPieces/ChessPiece.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+  protected abstract boolean validateMovement(ChessPiece[][] board, char xAxis, char yAxis);
+
+  public abstract char getClassSymbol();
+
+  public boolean isValidMovement(ChessPiece[][] board, char xAxis, char yAxis) {
+    if (!isValidPosition(xAxis, yAxis)) {
+      return false;
+    } else {
+      return validateMovement(board, xAxis, yAxis);
+    }
```
Comment (AlexisGlez) [inline Chess/ChessPieces/ChessPiece.java:23]: When you only have if's with return statements in your code, you don't need to use the else:
```
if (!isValidPosition(xAxis, yAxis)) {
  return false;
}

return validateMovement(board, xAxis, yAxis);
```

File: Chess/ChessPieces/ChessPiece.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+  }
+
+  public void setPosition(char xAxis, char yAxis) {
+    this.xAxis = xAxis;
+    this.yAxis = yAxis;
+  }
+
+  public boolean isWhite() {
+    return isWhite;
+  }
```
Comment (AlexisGlez) [inline Chess/ChessPieces/ChessPiece.java:40]: Bad practice to use setters and getters... why do you need this?

File: Chess/ChessPieces/ChessPiece.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+  protected boolean isLineWithClearPath(ChessPiece[][] board, char xAxis, char yAxis, int xDirection, int yDirection) {
+    xAxis += xDirection;
+    yAxis += yDirection;
+    while (this.xAxis != xAxis && this.yAxis != yAxis) {
+      int row = indexOfY(yAxis);
+      int col = indexOfX(xAxis);
+      if (board[row][col] != null)
+        return false;
```
Comment (AlexisGlez) [inline Chess/ChessPieces/ChessPiece.java:84]: Remember to use `{ }` to be consistent

File: Chess/ChessPieces/ChessPiece.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+  public static int indexOfX(char xAxis) {
+    return xAxis - 'a';
+  }
+
+  public static int indexOfY(char yAxis) {
+    return 7 - (yAxis - '1');
+  }
+
+  protected boolean isLineWithClearPath(ChessPiece[][] board, char xAxis, char yAxis, int xDirection, int yDirection) {
```
Comment (AlexisGlez) [inline Chess/ChessPieces/ChessPiece.java:77]: I still don't like this logic here :), remember me to give you more information offline

File: Chess/ChessPieces/Pieces/Bishop.java [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+public class Bishop extends ChessPiece {
+  public static char symbol = 'B';
+
+  public Bishop(char xAxis, char yAxis, boolean isWhite) {
+    super(xAxis, yAxis, isWhite);
+  }
+
+  protected boolean validateMovement(ChessPiece[][] board, char xAxis, char yAxis) {
+    // if distance in x does not equal distance in y it is not a valid movement
+    if (absoluteDifferenceX(xAxis) != absoluteDifferenceY(yAxis)) {
```
Comment (AlexisGlez) [inline Chess/ChessPieces/Pieces/Bishop.java:14]: It's weird to see  methods that exist in a parent class without the usage of `this`

File: Chess/ChessPieces/Pieces/Pawn.java [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+  }
+
+  protected boolean validateMovement(ChessPiece[][] board, char xAxis, char yAxis) {
+    if (this.xAxis == xAxis) {
+      return validatePawnVerticalMovement(board, xAxis, yAxis);
+
+    } else if (absoluteDifferenceX(xAxis) == 1) {
+      return validatePawnDiagonalMovement(board, xAxis, yAxis);
+
+    } else {
```
Comment (AlexisGlez) [inline Chess/ChessPieces/Pieces/Pawn.java:19]: Else is not necessary as you are already returning above

File: Chess/ChessPieces/Pieces/Pawn.java [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+      if (board[row - res][col] != null) {
+        return false;
+      }
+
+    } else if (absoluteDifferenceY(yAxis) == 2) {
+      int col = indexOfX(this.xAxis);
+
+      if (this.isWhite) {
+        return validateWhitePawnFirstMovement(board, col);
+      } else {
```
Comment (AlexisGlez) [inline Chess/ChessPieces/Pieces/Pawn.java:60]: Else not neccesary

File: Chess/ChessPieces/Pieces/Pawn.java [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+      return false;
+    }
+
+    return true;
+  }
+
+  private boolean validateWhitePawnFirstMovement(ChessPiece[][] board, int col) {
+    if (this.isWhite && this.yAxis == '2' && board[5][col] == null && board[4][col] == null) {
+      return true;
+    } else {
```
Comment (AlexisGlez) [inline Chess/ChessPieces/Pieces/Pawn.java:74]: Else not neccesary

File: Chess/ChessPieces/Pieces/Pawn.java [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+      return false;
+    }
+  }
+
+  private boolean validateBlackPawnFirstMovement(ChessPiece[][] board, int col) {
+    if (!this.isWhite && this.yAxis == '7' && board[2][col] == null && board[3][col] == null) {
+      return true;
+    } else {
```
Comment (AlexisGlez) [inline Chess/ChessPieces/Pieces/Pawn.java:82]: Else not necessary

File: DriverClass.java [hunk 10]
```diff
@@ -1,5 +1,22 @@
...
+    // int count = 0;
+    // for (char i = 'a'; i <= 'h'; i++) {
+    // for (char j = '1'; j <= '8'; j++) {
+    // System.out.println(i + "" + j + ": " + board[1][3].isValidMovement(board, i,
+    // j));
+    // if (board[1][3].isValidMovement(board, i, j))
+    // count++;
+    // }
+    // }
+    // System.out.println(count);
```
Comment (AlexisGlez) [inline DriverClass.java:20]: ?

**Review by AlexisGlez — COMMENTED** [review]
Code looks far more better than before 👍

File: Chess/ChessPieces/ChessPiece.java (hunk 2 above)
Comment (kevinglezgmz) [reply Chess/ChessPieces/ChessPiece.java:40]: To know if the piece is white or not and be able to associate it with player one or player two.

**Pull request merged**